server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GlobalOrdinalValuesSource.java

```diff
@@ -53,10 +75,51 @@ class GlobalOrdinalValuesSource extends SingleDimensionValuesSource<BytesRef> {
         int reverseMul
     ) {
         super(bigArrays, format, type, missingBucket, missingOrder, size, reverseMul);
+        this.uniqueValueCount = uniqueValueCount;
         this.docValuesFunc = docValuesFunc;
         this.values = bigArrays.newLongArray(Math.min(size, 100), false);
     }
 
+    /**
+     * Return the number of unique values of this source. Note that some unique
+     * values might not produce buckets if the query doesn't match documents
+     * that contain these values.
+     */
+    long getUniqueValueCount() {
+        return uniqueValueCount;
+    }
+
+    /**
+     * Return whether the source can be used for dynamic pruning.
+     */
+    boolean mayDynamicallyPrune(IndexReader indexReader) {
+        // If missing bucket is requested we have no way to tell lucene to efficiently match
+        // buckets in a range AND missing values.
+        if (missingBucket) {
+            return false;
+        }
+        // We also need to check if there are terms for the field in question.
+        // Some field types do not despite being global ordinals (e.g. IP addresses).
+        return hasTerms(indexReader);
+    }
+
+    private boolean hasTerms(IndexReader indexReader) {
+        List<LeafReaderContext> leaves = indexReader.leaves();
+        for (LeafReaderContext leaf : leaves) {
+            Terms terms;
+            try {
+                terms = leaf.reader().terms(fieldType.name());
+            } catch (IOException e) {
+                logger.error("failed to read terms; dynamic pruning is disabled", e);
+                return false;
```

Comment:
Getting an IOException here means that something bad happens, I would rather propagate it (possibly though an `UncheckedIOException` if we don't want to change method signatures)